include/fiction/technology/sidb_cluster_hierarchy.hpp

```diff
@@ -0,0 +1,1080 @@
+     */
+    uint64_t neg_count : 32;
+    /**
+     * Number of positive charges in the cluster charge state (32 available bits).
+     */
+    uint64_t pos_count : 32;
+
+    mutable std::vector<sidb_cluster_state_composition> compositions{};
+    /**
+     * Default constructor, creates a cluster charge state without any negative and positive charges.
+     */
+    explicit sidb_cluster_charge_state() noexcept : neg_count{0}, pos_count{0} {}
+    /**
+     * Constructor for a charge space element of a singleton cluster. It has a single composition, which is a cluster
+     * state of the singleton cluster and the singleton multiset charge configuration itself.
+     *
+     * @param singleton Singleton cluster to put in the compositions of this cluster charge state.
+     * @param cs Charge state to lift to a singleton multiset charge configuration.
+     * @param loc_ext_pot The local external potential at the SiDB in the singleton cluster. Specifically, this is the
+     * sum of the local defect potential and the local external potential.
+     */
+    explicit sidb_cluster_charge_state(const sidb_cluster_ptr& singleton, const sidb_charge_state cs,
+                                       const double loc_ext_pot) noexcept :
+            neg_count{static_cast<decltype(neg_count)>(cs == sidb_charge_state::NEGATIVE)},
+            pos_count{static_cast<decltype(pos_count)>(cs == sidb_charge_state::POSITIVE)},
+            compositions{{sidb_cluster_state{singleton, static_cast<uint64_t>(*this)}}}
+    {
+        compositions.front().front().set_pot_bounds(get_singleton_sidb_ix(singleton), loc_ext_pot, loc_ext_pot);
+    }
+    /**
+     * Constructor for cluster charge state given a multiset charge configuration represented in its compressed form. It
+     * allows the compressed form to be lifted to the full type to facilitate equality checks.
+     *
+     * @param m The multiset charge configuration to create a cluster charge state of.
+     */
+    explicit sidb_cluster_charge_state(const uint64_t m) noexcept :
+            neg_count{static_cast<decltype(neg_count)>(m >> 32ull)},
+            pos_count{static_cast<decltype(pos_count)>((m << 32ull) >> 32ull)}
+    {}
+    /**
+     * Explicit instructions for the compiler on how to cast a cluster charge state to an 64 bit unsigned integer.
+     *
+     * @return The 64 bit unsigned integer representing the compressed form of the cluster charge state.
+     */
+    explicit constexpr operator uint64_t() const noexcept
+    {
+        return (static_cast<uint64_t>(neg_count) << 32ull) | pos_count;
+    }
+    /**
+     * Modifier of the cluster charge state, adding a single charge state.
+     *
+     * @param cs The charge state to add.
+     */
+    constexpr void add_charge(const sidb_charge_state cs) noexcept
+    {
+        switch (cs)
+        {
+            case sidb_charge_state::NEGATIVE: neg_count++; break;
```

Comment:
```suggestion
            case sidb_charge_state::NEGATIVE: 
            {
                neg_count++; 
                break;
            }
```
Please do the same for the others.